Changes in module 'nf-core/minimap2/align'
--- modules/nf-core/minimap2/align/main.nf
+++ modules/nf-core/minimap2/align/main.nf
@@ -1,6 +1,5 @@
 process MINIMAP2_ALIGN {
     tag "$meta.id"
-    label 'process_high'
 
     // Note: the versions here need to match the versions used in the mulled container below and minimap2/index
     conda "${moduleDir}/environment.yml"

************************************************************
