# sanger-tol/blobtoolkit: Usage

> _Documentation of pipeline parameters is generated automatically from the pipeline schema and can no longer be found in markdown files._

## Introduction

<!-- Add documentation about anything specific to running your pipeline. For general topics, please point to (and add to) the main nf-core website. -->

## Samplesheet input

You will need to create a samplesheet with information about the samples you would like to analyse before running the pipeline. Use this parameter to specify its location. It has to be a comma-separated file with 4 columns, and a header row as shown in the examples below.

```bash
--input '[path to samplesheet file]'
```

### Multiple runs of the same sample

The `sample` identifiers have to be the same when you have re-sequenced the same sample more than once e.g. to increase sequencing depth. The pipeline will concatenate the raw reads before performing any downstream analysis. Below is an example for the same sample sequenced across 3 lanes:

```console
sample,datatype,datafile,library_layout
sample1,hic,hic.cram,PAIRED
sample2,illumina,illumina.cram,PAIRED
sample2,illumina,illumina.cram,PAIRED
```

### Full samplesheet

The samplesheet can have as many columns as you desire, however, there is a strict requirement for the first 4 columns to match those defined in the table below.

A final samplesheet file may look something like the one below.

```console
sample,datatype,datafile,library_layout
sample1,hic,hic.cram,PAIRED
sample2,illumina,illumina.cram,PAIRED
sample3,ont,ont.cram,SINGLE
```

| Column           | Description                                                                                                                                                                           |
| ---------------- | ------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- |
| `sample`         | Custom sample name. This entry will be identical for multiple sequencing libraries/runs from the same sample. Spaces in sample names are automatically converted to underscores (\_). |
| `datatype`       | Type of sequencing data. Must be one of `hic`, `illumina`, `pacbio`, `pacbio_clr` or `ont`.                                                                                           |
| `datafile`       | Full path to read data file.                                                                                                                                                          |
| `library_layout` | Layout of the library. Must be one of `SINGLE`, `PAIRED`.                                                                                                                             |

An [example samplesheet](assets/test/samplesheet.csv) has been provided with the pipeline.

### Support for [nf-core/fetchngs](https://nf-co.re/fetchngs)

The pipeline can also accept a samplesheet generated by the [nf-core/fetchngs](https://nf-co.re/fetchngs) pipeline (tested with version 1.11.0).
The pipeline then needs the `--fetchngs_samplesheet true` option _and_ `--align true`, since the data files would all be unaligned.

### Support for pre-computed `BUSCO` outputs

The pipeline may be optionally run with a set of pre-computed [`BUSCO`](https://busco.ezlab.org) runs, provided using the `--precomputed_busco` parameter. These can be provided as either a directory path, or a `.tar.gz` compressed archive. The contents should be each `run_` output directory (directly from `BUSCO`) named as `run_[odb_dabasase_name]`:

```
GCA_922984935.2_busco_output/
├── run_archaea_odb10
├── run_bacteria_odb10
├── run_carnivora_odb10
├── run_eukaryota_odb10
├── run_eutheria_odb10
├── run_laurasiatheria_odb10
├── run_mammalia_odb10
├── run_metazoa_odb10
├── run_tetrapoda_odb10
└── run_vertebrata_odb10
```

The pipeline minimally requires outputs for the 'basal' lineages (archaea, eukaryota, and bacteria) -- any of these which are not present in the pre-computed outputs will be automatically detected and run.

## Database parameters

Configure access to your local databases with the `--busco`, `--blastp`, `--blastx`, `--blastn`, and `--taxdump` parameters.

Note that `--busco` refers to the download path of _all_ lineages.
Then, when explicitly selecting the lineages to run the pipeline on,
provide the names of these lineages _with_ their `_odb10` suffix as a comma-separated string.
For instance:

```bash
--busco path-to-databases/busco/ --busco_lineages vertebrata_odb10,bacteria_odb10,fungi_odb10
```

### Getting databases ready for the pipeline

The BlobToolKit pipeline can be run in many different ways. The default way requires access to several databases:

1. [NCBI taxdump database](https://www.ncbi.nlm.nih.gov/taxonomy)
2. [NCBI nucleotide BLAST database](https://blast.ncbi.nlm.nih.gov/doc/blast-help/downloadblastdata.html#databases)
3. [UniProt reference proteomes database](https://www.uniprot.org)
4. [BUSCO database](https://busco.ezlab.org)

It is a good idea to put a date suffix for each database location so you know at a glance whether you are using the latest version. We are using the `YYYY_MM` format as we do not expect the databases to be updated more frequently than once a month. However, feel free to use `DATE=YYYY_MM_DD` or a different format if you prefer.

Note that all input databases may be optionally passed directly to the pipeline compressed as `.tar.gz`, and the pipeline will handle decompression.
The instructions below show how to build each input database in _two_ forms: decompressed _and_ compressed. You may not need to do both. Select the one that is most appropriate for how you want to use the pipeline.

#### 1. NCBI taxdump database

Create the database directory, retrieve and decompress the NCBI taxonomy:

```bash
DATE=2024_10
TAXDUMP=/path/to/databases/taxdump_${DATE}
TAXDUMP_TAR=/path/to/databases/taxdump_${DATE}.tar.gz
mkdir -p "$TAXDUMP"
curl -L ftp://ftp.ncbi.nih.gov/pub/taxonomy/new_taxdump/new_taxdump.tar.gz -o $TAXDUMP_TAR
tar -xzf $TAXDUMP_TAR -C "$TAXDUMP"
```

#### 2. NCBI nucleotide BLAST database

Create the database directory and move into the directory:

```bash
DATE=2024_10
NT=/path/to/databases/nt_${DATE}
NT_TAR=/path/to/databases/nt_${DATE}.tar.gz
mkdir -p $NT
cd $NT
```

Retrieve the NCBI blast nt database (version 5) files and tar gunzip them.
`wget` and the use of the FTP protocol are necessary to resolve the wildcard `nt.???.tar.gz`.
We are using the `&&` syntax to ensure that each command completes without error before the next one is run:

```bash
wget "ftp://ftp.ncbi.nlm.nih.gov/blast/db/v5/nt.???.tar.gz" -P $NT/ &&
for file in $NT/*.tar.gz; do
    tar xf $file -C $NT && rm $file;
done

wget "https://ftp.ncbi.nlm.nih.gov/blast/db/v5/taxdb.tar.gz" &&
tar xf taxdb.tar.gz -C $NT &&
rm taxdb.tar.gz

# Compress and cleanup
cd ..
tar -cvzf $NT_TAR $NT
rm -r $NT
```

#### 3. UniProt reference proteomes database

You need [diamond blast](https://github.com/bbuchfink/diamond) installed for this step.
The easiest way is probably to install their [pre-compiled release](https://github.com/bbuchfink/diamond/releases).
Make sure you have the latest version of Diamond (>2.x.x) otherwise the `--taxonnames` argument may not work.

Create the database directory and move into the directory:

```bash
DATE=2024_10
UNIPROT=/path/to/databases/uniprot_${DATE}
UNIPROT_TAR=/path/to/databases/uniprot_${DATE}.tar.gz
mkdir -p $UNIPROT
cd $UNIPROT
```

The UniProt `Refseq_Proteomes_YYYY_MM.tar.gz` file is very large (close to 200 GB) and will take a long time to download.
The command below looks complex because it needs to get around the problem of using wildcards with wget and curl.

```bash
EBI_URL=ftp.ebi.ac.uk/pub/databases/uniprot/current_release/knowledgebase/reference_proteomes/
mkdir extract
curl -L $EBI_URL/$(curl -vs $EBI_URL 2>&1 | awk '/tar.gz/ {print $9}') | \
  tar -xzf - -C extract

# Create a single fasta file with all the fasta files from each subdirectory:
find extract -type f -name '*.fasta.gz' ! -name '*_DNA.fasta.gz' ! -name '*_additional.fasta.gz' -exec cat '{}' '+' > reference_proteomes.fasta.gz

# create the accession-to-taxid map for all reference proteome sequences:
find extract -type f -name '*.idmapping.gz' -exec zcat {} + | \
  awk 'BEGIN {OFS="\t"; print "accession", "accession.version", "taxid", "gi"} $2=="NCBI_TaxID" {print $1, $1, $3, 0}' > reference_proteomes.taxid_map

# create the taxon aware diamond blast database
diamond makedb -p 16 --in reference_proteomes.fasta.gz --taxonmap reference_proteomes.taxid_map --taxonnodes $TAXDUMP/nodes.dmp --taxonnames $TAXDUMP/names.dmp -d reference_proteomes.dmnd

# clean up
mv extract/{README,STATS} .
rm -r extract
rm -r $TAXDUMP

# Compress final database and cleanup
cd ..
tar -cvzf $UNIPROT_TAR $UNIPROT
rm -r $UNIPROT
```

#### 4. BUSCO databases

Create the database directory and move into the directory:

```bash
DATE=2024_10
BUSCO=/path/to/databases/busco_${DATE}
BUSCO_TAR=/path/to/databases/busco_${DATE}.tar.gz
mkdir -p $BUSCO
cd $BUSCO
```

Download BUSCO data and lineages to allow BUSCO to run in offline mode:

```bash
wget -r -nH https://busco-data.ezlab.org/v5/data/
# the trailing slash after data is important. Otherwise wget doesn't get the subdirectories

# tar gunzip all folders that have been stored as tar.gz, in the same parent directories as where they were stored:
find v5/data -name "*.tar.gz" | while read -r TAR; do tar -C `dirname $TAR` -xzf $TAR; done
```

If you have [GNU parallel](https://www.gnu.org/software/parallel/) installed, you can also use the command below which will run faster as it will run the decompression commands in parallel:

```bash
find v5/data -name "*.tar.gz" | parallel "cd {//}; tar -xzf {/}"
```

Finally re-compress and cleanup the files:

```bash
tar -cvzf $BUSCO_TAR $BUSCO
rm -r $BUSCO
```

## Changes from Snakemake to Nextflow

### Commands

Snakemake

```bash
# Public Assemblies
run_btk_pipeline.sh GCA_ACCESSION

# Draft Assemblies
blobtoolkit-pipeline run --config YAML --threads INT --workdir DIR
```

Nextflow

```bash
# Public Assemblies
nextflow run sanger-tol/blobtoolkit --input SAMPLESHEET --fasta GENOME –-accession GCA_ACCESSION --taxon TAXON_ID --taxdump TAXDUMP_DB --blastp DMND_db --blastn BLASTN_DB --blastx BLASTX_DB

# Draft Assemblies
nextflow run sanger-tol/blobtoolkit --input SAMPLESHEET --fasta GENOME --taxon TAXON_ID --taxdump TAXDUMP_DB --blastp DMND_db --blastn BLASTN_DB --blastx BLASTX_DB
```

The Nextflow pipeline does not support taking input from the Yaml files of the Snakemake pipeline.
Instead, Nextflow has a uniform way of setting input parameters on the command-line or via a JSON / Yaml file,
see <https://training.nextflow.io/basic_training/config/> for some examples.

### Subworkflows

Here is a full list of snakemake subworkflows and their Nextflow couterparts:

- **`minimap.smk`**
  - Implemented as [`minimap_alignment.nf`](subworkflows/local/minimap_alignment.nf).
  - Optimised alignment is done using the [sanger-tol/readmapping](https://github.com/sanger-tol/readmapping) pipeline.
- **`windowmasker.smk`**
  - Implemented as part of [`prepare_genome.nf`](subworkflows/local/prepare_genome.nf).
  - Genomes downloaded by [sanger-tol/insdcdownload](https://github.com/sanger-tol/insdcdownload) are already masked.
- **`chunk_stats.smk`**
  - Modified implementation as part of [`coverage_stats.nf`](subworkflows/local/coverage_stats.nf).
  - BED file and additional statistics calculated using [`fasta_windows`](https://github.com/tolkit/fasta_windows).
- **`busco.smk`**
  - Implemented as [`busco_diamond_blastp.nf`](subworkflows/local/busco_diamond_blastp.nf).
- **`cov_stats.smk`**
  - Implemented as part of [`coverage_stats.nf`](subworkflows/local/coverage_stats.nf).
  - Combining the various tsv files is done in subworkflow [`collate_stats.nf`](subworkflows/local/collate_stats.nf).
- **`window_stats.smk`**
  - Implemented as part of [`collate_stats.nf`](subworkflows/local/collate_stats.nf).
- **`diamond_blastp.smk`**
  - Implemented as [`busco_diamond_blastp.nf`](subworkflows/local/busco_diamond_blastp.nf).
- **`diamond.smk`**
  - Implemented as [`run_blastx.nf`](subworkflows/local/run_blastx.nf).
- **`blastn.smk`**
  - Implemented as [`run_blastn.nf`](subworkflows/local/run_blastn.nf).
- **`blobtools.smk`**
  - Implemented as [`blobtools.nf`](subworkflows/local/blobtools.nf).
- **`view.smk`**
  - Implemented as [`view.nf`](subworkflows/local/view.nf).

### Software dependencies

List of tools for any given dataset can be fetched from the API, for example https://blobtoolkit.genomehubs.org/api/v1/dataset/id/CAJEUD01.1/settings/software_versions.

| Dependency        | Snakemake | Nextflow |
| ----------------- | --------- | -------- |
| blobtoolkit       | 4.3.2     | 4.4.4    |
| blast             | 2.12.0    | 2.14.1   |
| blobtk            | 0.5.0     | 0.5.1    |
| busco             | 5.3.2     | 5.5.0    |
| diamond           | 2.0.15    | 2.1.8    |
| fasta_windows     |           | 0.2.4    |
| minimap2          | 2.24      | 2.24     |
| ncbi-datasets-cli | 14.1.0    |          |
| nextflow          |           | 23.10.0  |
| python            | 3.9.13    | 3.12.0   |
| samtools          | 1.15.1    | 1.19.2   |
| seqtk             | 1.3       | 1.4      |
| snakemake         | 7.19.1    |          |
| windowmasker      | 2.12.0    | 2.14.0   |

> **NB:** Dependency has been **added** if only the Nextflow version information is present.
> **NB:** Dependency has been **removed** if only the Snakemake version information is present.
> **NB:** Dependency has been **updated** if bothe the Snakemake and Nextflow version information is present.

## Running the pipeline

The typical command for running the pipeline is as follows:

```bash
nextflow run sanger-tol/blobtoolkit --input samplesheet.csv --outdir <OUTDIR> --fasta genome.fasta -profile docker –-accession GCA_accession --taxon "species name" --taxdump /path/to/taxdump --blastp /path/to/buscogenes.dmnd --blastn /path/to/blastn.nt --blastx /path/to/buscoregions.dmnd
```

This will launch the pipeline with the `docker` configuration profile. See below for more information about profiles.

Note that the pipeline will create the following files in your working directory:

```bash
work                # Directory containing the nextflow working files
<OUTDIR>            # Finished results in specified location (defined with --outdir)
.nextflow_log       # Log file from Nextflow
# Other nextflow hidden files, eg. history of pipeline runs and old logs.
```

If you wish to repeatedly use the same parameters for multiple runs, rather than specifying each flag in the command, you can specify these in a params file.

Pipeline settings can be provided in a `yaml` or `json` file via `-params-file <file>`.

> [!WARNING]
> Do not use `-c <file>` to specify parameters as this will result in errors. Custom config files specified with `-c` must only be used for [tuning process resource specifications](https://nf-co.re/docs/usage/configuration#tuning-workflow-resources), other infrastructural tweaks (such as output directories), or module arguments (args).

The above pipeline run specified with a params file in yaml format:

```bash
nextflow run sanger-tol/blobtoolkit -profile docker -params-file params.yaml
```

with:

```yaml title="params.yaml"
input: './samplesheet.csv'
outdir: './results/'
<...>
```

You can also generate such `YAML`/`JSON` files via [nf-core/launch](https://nf-co.re/launch).

### Updating the pipeline

When you run the above command, Nextflow automatically pulls the pipeline code from GitHub and stores it as a cached version. When running the pipeline after this, it will always use the cached version if available - even if the pipeline has been updated since. To make sure that you're running the latest version of the pipeline, make sure that you regularly update the cached version of the pipeline:

```bash
nextflow pull sanger-tol/blobtoolkit
```

### Reproducibility

It is a good idea to specify the pipeline version when running the pipeline on your data. This ensures that a specific version of the pipeline code and software are used when you run your pipeline. If you keep using the same tag, you'll be running the same version of the pipeline, even if there have been changes to the code since.

First, go to the [sanger-tol/blobtoolkit releases page](https://github.com/sanger-tol/blobtoolkit/releases) and find the latest pipeline version - numeric only (eg. `1.3.1`). Then specify this when running the pipeline with `-r` (one hyphen) - eg. `-r 1.3.1`. Of course, you can switch to another version by changing the number after the `-r` flag.

This version number will be logged in reports when you run the pipeline, so that you'll know what you used when you look back in the future. For example, at the bottom of the MultiQC reports.

To further assist in reproducibility, you can use share and reuse [parameter files](#running-the-pipeline) to repeat pipeline runs with the same settings without having to write out a command with every single parameter.

> [!TIP]
> If you wish to share such profile (such as upload as supplementary material for academic publications), make sure to NOT include cluster specific paths to files, nor institutional specific profiles.

## Core Nextflow arguments

> [!NOTE]
> These options are part of Nextflow and use a _single_ hyphen (pipeline parameters use a double-hyphen)

### `-profile`

Use this parameter to choose a configuration profile. Profiles can give configuration presets for different compute environments.

Several generic profiles are bundled with the pipeline which instruct the pipeline to use software packaged using different methods (Docker, Singularity, Podman, Shifter, Charliecloud, Apptainer, Conda) - see below.

> [!IMPORTANT]
> We highly recommend the use of Docker or Singularity containers for full pipeline reproducibility, however when this is not possible, Conda is also supported.

The pipeline also dynamically loads configurations from [https://github.com/nf-core/configs](https://github.com/nf-core/configs) when it runs, making multiple config profiles for various institutional clusters available at run time. For more information and to check if your system is supported, please see the [nf-core/configs documentation](https://github.com/nf-core/configs#documentation).

Note that multiple profiles can be loaded, for example: `-profile test,docker` - the order of arguments is important!
They are loaded in sequence, so later profiles can overwrite earlier profiles.

If `-profile` is not specified, the pipeline will run locally and expect all software to be installed and available on the `PATH`. This is _not_ recommended, since it can lead to different results on different machines dependent on the computer environment.

- `test`
  - A profile with a complete configuration for automated testing
  - Includes links to test data so needs no other parameters
- `docker`
  - A generic configuration profile to be used with [Docker](https://docker.com/)
- `singularity`
  - A generic configuration profile to be used with [Singularity](https://sylabs.io/docs/)
- `podman`
  - A generic configuration profile to be used with [Podman](https://podman.io/)
- `shifter`
  - A generic configuration profile to be used with [Shifter](https://nersc.gitlab.io/development/shifter/how-to-use/)
- `charliecloud`
  - A generic configuration profile to be used with [Charliecloud](https://hpc.github.io/charliecloud/)
- `apptainer`
  - A generic configuration profile to be used with [Apptainer](https://apptainer.org/)
- `wave`
  - A generic configuration profile to enable [Wave](https://seqera.io/wave/) containers. Use together with one of the above (requires Nextflow ` 24.03.0-edge` or later).
- `conda`
  - A generic configuration profile to be used with [Conda](https://conda.io/docs/). Please only use Conda as a last resort i.e. when it's not possible to run the pipeline with Docker, Singularity, Podman, Shifter, Charliecloud, or Apptainer.

### `-resume`

Specify this when restarting a pipeline. Nextflow will use cached results from any pipeline steps where the inputs are the same, continuing from where it got to previously. For input to be considered the same, not only the names must be identical but the files' contents as well. For more info about this parameter, see [this blog post](https://www.nextflow.io/blog/2019/demystifying-nextflow-resume.html).

You can also supply a run name to resume a specific run: `-resume [run-name]`. Use the `nextflow log` command to show previous run names.

### `-c`

Specify the path to a specific config file (this is a core Nextflow command). See the [nf-core website documentation](https://nf-co.re/usage/configuration) for more information.

## Custom configuration

### Resource requests

Whilst the default requirements set within the pipeline will hopefully work for most people and with most input data, you may find that you want to customise the compute resources that the pipeline requests. Each step in the pipeline has a default set of requirements for number of CPUs, memory and time. For most of the pipeline steps, if the job exits with any of the error codes specified [here](https://github.com/nf-core/rnaseq/blob/4c27ef5610c87db00c3c5a3eed10b1d161abf575/conf/base.config#L18) it will automatically be resubmitted with higher resources request (2 x original, then 3 x original). If it still fails after the third attempt then the pipeline execution is stopped.

To change the resource requests, please see the [max resources](https://nf-co.re/docs/usage/configuration#max-resources) and [tuning workflow resources](https://nf-co.re/docs/usage/configuration#tuning-workflow-resources) section of the nf-core website.

### Custom Containers

In some cases, you may wish to change the container or conda environment used by a pipeline steps for a particular tool. By default, nf-core pipelines use containers and software from the [biocontainers](https://biocontainers.pro/) or [bioconda](https://bioconda.github.io/) projects. However, in some cases the pipeline specified version maybe out of date.

To use a different container from the default container or conda environment specified in a pipeline, please see the [updating tool versions](https://nf-co.re/docs/usage/configuration#updating-tool-versions) section of the nf-core website.

### Custom Tool Arguments

A pipeline might not always support every possible argument or option of a particular tool used in pipeline. Fortunately, nf-core pipelines provide some freedom to users to insert additional parameters that the pipeline does not include by default.

To learn how to provide additional arguments to a particular tool of the pipeline, please see the [customising tool arguments](https://nf-co.re/docs/usage/configuration#customising-tool-arguments) section of the nf-core website.

### nf-core/configs

In most cases, you will only need to create a custom config as a one-off but if you and others within your organisation are likely to be running nf-core pipelines regularly and need to use the same settings regularly it may be a good idea to request that your custom config file is uploaded to the `nf-core/configs` git repository. Before you do this please can you test that the config file works with your pipeline of choice using the `-c` parameter. You can then create a pull request to the `nf-core/configs` repository with the addition of your config file, associated documentation file (see examples in [`nf-core/configs/docs`](https://github.com/nf-core/configs/tree/master/docs)), and amending [`nfcore_custom.config`](https://github.com/nf-core/configs/blob/master/nfcore_custom.config) to include your custom profile.

See the main [Nextflow documentation](https://www.nextflow.io/docs/latest/config.html) for more information about creating your own configuration files.

If you have any questions or issues please send us a message on [Slack](https://nf-co.re/join/slack) on the [`#configs` channel](https://nfcore.slack.com/channels/configs).

## Running in the background

Nextflow handles job submissions and supervises the running jobs. The Nextflow process must run until the pipeline is finished.

The Nextflow `-bg` flag launches Nextflow in the background, detached from your terminal so that the workflow does not stop if you log out of your session. The logs are saved to a file.

Alternatively, you can use `screen` / `tmux` or similar tool to create a detached session which you can log back into at a later time.
Some HPC setups also allow you to run nextflow within a cluster job submitted your job scheduler (from where it submits more jobs).

## Nextflow memory requirements

In some cases, the Nextflow Java virtual machines can start to request a large amount of memory.
We recommend adding the following line to your environment to limit this (typically in `~/.bashrc` or `~./bash_profile`):

```bash
NXF_OPTS='-Xms1g -Xmx4g'
```
